internal/engine/portforward/subscriber.go

```diff
@@ -103,25 +111,28 @@ func (s *Subscriber) OnChange(ctx context.Context, st store.RStore, _ store.Chan
 	toStart, toShutdown := s.diff(ctx, st)
 	for _, entry := range toShutdown {
 		entry.cancel()
+
+		st.Dispatch(NewPortForwardDeleteAction(entry.Name))
+		// TODO(maia): delete PF object in API
 	}
 
 	for _, entry := range toStart {
 		// Treat port-forwarding errors as part of the pod log
 		ctx := logger.CtxWithLogHandler(entry.ctx, PodLogActionWriter{
 			Store:        st,
-			PodID:        entry.podID,
-			ManifestName: entry.name,
+			PodID:        k8s.PodID(entry.Spec.PodName),
+			ManifestName: model.ManifestName(entry.ObjectMeta.Annotations[v1alpha1.AnnotationManifest]),
 		})
 
-		for _, forward := range entry.forwards {
-			entry := entry
-			forward := forward
-			go s.startPortForwardLoop(ctx, entry, forward)
-		}
+		go s.startPortForwardLoop(ctx, entry)
+
+		st.Dispatch(NewPortForwardCreateAction(entry.PortForward))
```

Comment:
i'm pretty sure you're going to run into problems with this approach.

usually the data flow works like
subscriber reads EngineState -> creates API objects
reconciler reads API Objects -> sets up external work

The migration path we recommend tries to get the data flow working first

I think it's going to be really hard to start with the opposite data flow, and then try to reverse it

Reply:
(this is what Milas and I both originally tried to do and both of us hit a wall :grimacing: )

Reply:
> I think it's going to be really hard to start with the opposite data flow, and then try to reverse it

how do you figure this is the opposite data flow? This is me writing the API objects to the engine state. I think I'm missing something here...

Reply:
@milas has done more of these than I have at this point and can probably give better guidance / pointers on next steps (or if he sees possible paths from here)

by "opposite", i mean this code creates the PortForward entry + loop first, then sends the PortForward API object to the EngineState. Usually, you'd go in the opposite direction - send the PortForward API object to the engine state, then read it off the engine state to set up the port forward in a separate reconciler.

Reply:
oh yes that is my plan for the next PR -- i did this instead of a bigger PR where I both write the new object to engine state AND move over all the logic for running the port forwards to a different controller.

are you saying you'd rather see the latter type of PR, or that there's a different and less foot-gun-y way to do this incrementally?

Reply:
talked to milas offline, I'm going to try again with an incremental approach, but one that doesn't risk infinite loops by dispatching events.